File: api/middleware/rateLimit.js
const { SAFETY_LIMITS } = require('../config/constants');
const { getClientIp } = require('../utils/ip');

const rateLimitStore = new Map();

function checkRateLimit(ip) {
  const now = Date.now();
  const windowStart = now - SAFETY_LIMITS.rateLimitWindowMs;

  if (!rateLimitStore.has(ip)) {
    rateLimitStore.set(ip, []);
  }

  const requests = rateLimitStore.get(ip).filter(t => t > windowStart);
  rateLimitStore.set(ip, requests);

  if (requests.length >= SAFETY_LIMITS.rateLimitMaxRequests) {
    return { allowed: false, remaining: 0, resetIn: Math.ceil((requests[0] + SAFETY_LIMITS.rateLimitWindowMs - now) / 1000) };
  }

  requests.push(now);
  return { allowed: true, remaining: SAFETY_LIMITS.rateLimitMaxRequests - requests.length };
}

function rateLimitMiddleware(req, res, next) {
  const ip = getClientIp(req);
  const result = checkRateLimit(ip);

  res.setHeader('X-RateLimit-Limit', SAFETY_LIMITS.rateLimitMaxRequests);
  res.setHeader('X-RateLimit-Remaining', result.remaining);

  if (!result.allowed) {
    res.setHeader('X-RateLimit-Reset', result.resetIn);
    return res.status(429).json({
      error: 'Too many requests. Please slow down.',
      resetIn: result.resetIn
    });
  }

  next();
}

function startRateLimitCleanup() {
  setInterval(() => {
    const now = Date.now();
    const windowStart = now - SAFETY_LIMITS.rateLimitWindowMs;
    for (const [ip, requests] of rateLimitStore.entries()) {
      const valid = requests.filter(t => t > windowStart);
      if (valid.length === 0) {
        rateLimitStore.delete(ip);
      } else {
        rateLimitStore.set(ip, valid);
      }
    }
  }, 60000);
}

module.exports = {
  checkRateLimit,
  rateLimitMiddleware,
  startRateLimitCleanup
};

File: internal/middleware/ratelimit.go
package middleware

import (
	"encoding/json"
	"fmt"
	"net/http"
	"sync"
	"time"

	"github.com/coah80/yoink/internal/config"
	"github.com/coah80/yoink/internal/util"
)

var (
	rateLimitStore = make(map[string][]time.Time)
	rateLimitMu    sync.Mutex
)

func RateLimit(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ip := util.GetClientIP(r)
		allowed, remaining, resetIn := checkRateLimit(ip)

		w.Header().Set("X-RateLimit-Limit", fmt.Sprintf("%d", config.RateLimitMax))
		w.Header().Set("X-RateLimit-Remaining", fmt.Sprintf("%d", remaining))

		if !allowed {
			w.Header().Set("X-RateLimit-Reset", fmt.Sprintf("%d", resetIn))
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(429)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"error":   "Too many requests. Please slow down.",
				"resetIn": resetIn,
			})
			return
		}

		next.ServeHTTP(w, r)
	})
}

const maxRateLimitEntries = 100000

func checkRateLimit(ip string) (allowed bool, remaining int, resetIn int) {
	rateLimitMu.Lock()
	defer rateLimitMu.Unlock()

	now := time.Now()
	windowStart := now.Add(-config.RateLimitWindow)

	requests := rateLimitStore[ip]
	filtered := requests[:0]
	for _, t := range requests {
		if t.After(windowStart) {
			filtered = append(filtered, t)
		}
	}

	if len(filtered) >= config.RateLimitMax {
		resetSec := int(filtered[0].Add(config.RateLimitWindow).Sub(now).Seconds()) + 1
		rateLimitStore[ip] = filtered
		return false, 0, resetSec
	}

	if len(rateLimitStore) >= maxRateLimitEntries {
		rateLimitStore[ip] = filtered
		return false, 0, 60
	}

	filtered = append(filtered, now)
	rateLimitStore[ip] = filtered
	return true, config.RateLimitMax - len(filtered), 0
}

func StartRateLimitCleanup() {
	go func() {
		ticker := time.NewTicker(60 * time.Second)
		for range ticker.C {
			rateLimitMu.Lock()
			now := time.Now()
			windowStart := now.Add(-config.RateLimitWindow)
			for ip, requests := range rateLimitStore {
				filtered := requests[:0]
				for _, t := range requests {
					if t.After(windowStart) {
						filtered = append(filtered, t)
					}
				}
				if len(filtered) == 0 {
					delete(rateLimitStore, ip)
				} else {
					rateLimitStore[ip] = filtered
				}
			}
			rateLimitMu.Unlock()
		}
	}()
}
